mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingDecay.scala

```diff
@@ -0,0 +1,57 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.mllib.regression
+
+import org.apache.spark.Logging
+
+trait StreamingDecay {
+  def getDiscount(numNewDataPoints: Long): Double
+}
+
+private[mllib] trait StreamingDecaySetter[T <: StreamingDecaySetter[T]] extends Logging {
```

Comment:
"Does the code not work when you replace T with self.type?"

I guess not? For example, 
``` scala
trait Setter { def set: this.type = this}
class Apple extends Setter
val a = new Apple()
a.set
```
The return type of `a.set` is `a.type`, not `Apple`. Do I answer your question?

"Why do we need F-bounded polymorphism here?"
I agree with you that this is not needed here. Originally I included this as an extra level of type checking. But since I have `self: T=>` in the next line, I don't think we need it any more. I will remove it in the next push to the PR.